lib/ukdebug/README.md

```diff
@@ -0,0 +1,201 @@
+machine running Unikraft can be a local VM or a physically remote machine.
+
+```plain
+  Your machine (any OS)                Other machine (Unikraft)
+  ┌─────────────┐                         ┌─────────────┐
+  │             │    Serial connection    │             │
+  │  GDB Host ◄─┼─────────────────────────┼─► GDB Stub  │
+  │             │                         │             │
+  └─────────────┘                         └─────────────┘
+```
+
+Unikraft's GDB stub lives in `gdbstub.c` with architecture-specific code in
+`arch/$(UK_ARCH)/gdbsup.c`.
+
+The stub runs after every debug trap (`UKARCH_TRAP_DEBUG`) meaning the stub
+executes in an exception context. There is a loop inside the stub that waits to
+receive packets from the host and act upon them. Once the stub is finished
+handling a packet, it will send the response to the host and wait for the next
+packet. Some packets instruct the stub to leave the loop and resume execution
+of the kernel.
+
+The GDB stub uses a serial console device (`struct uk_console`) to communicate
+with the host. This device is referred to as the "backing console" because it
+backs the connection to the host. When the stub is initially set up, it
+retrieves a reference to the console device for the backing console from
+`ukconsole` and clears the `UK_CONSOLE_FLAG_STDIN` and `UK_CONSOLE_FLAG_STDOUT`
+flags for that device.
+
+Clearing the flags gives the GDB stub exclusive access to the backing console.
+If the GDB stub didn't have exclusive access to the backing console, any kernel
+output would interfere with the GDB communications and potentially break the
+connection. However, even on systems that only feature a single console, we
+would still like to see kernel output. For this reason, the GDB stub has a mode
+to share the backing console between the GDB communications and kernel output.
+This is done by re-routing all kernel output and sending it to the GDB host
+wrapped in a special packet. The GDB host then prints the output that it
+receives to `stdout`, inline with the normal output from the GDB CLI.
+
+### Kernel parameters
+
+Two kernel parameters control the backing console and the mode of the GDB stub.
+
+- `debug.gdbcon`: This is the console ID of the console device from `ukconsole`
+  that should be used as the backing console; it's a 16-bit unsigned integer.
+  If there's only a single console on the system, its ID is `0`, so
+  `debug.gdbcon=0` should be used to select it. The GDB stub is only enabled at
+  runtime if `debug.gdbcon` is set.
+- `debug.gdbmode`: This is a tristate value controlling the mode of the GDB
+  stub. Its type is a signed integer. The three states for the tristate are
+  "shared", "not shared", and "auto". Any negative value is interpreted to mean
+  "auto", a value of zero is interpreted to mean "not shared" and a positive
+  value is interpreted as "shared". "Shared" mode is useful when the system
```

Comment:
> "Shared" mode is useful when the system only has a single console, its operation was described above. 

"The "shared" mode...".

> If you don't include this kernel parameter, "auto" will be used.

You can skip that by adopting the sentece in comment above that states "The default is "auto"."